examples/cifar10_deepspeed_te.py

```diff
@@ -0,0 +1,406 @@
+#     the num_worker of torch.utils.data.DataLoader() to 0.
+
+transform = transforms.Compose([transforms.ToTensor(), transforms.Normalize((0.5, 0.5, 0.5), (0.5, 0.5, 0.5))])
+
+if torch.distributed.get_rank() != 0:
+    # might be downloading cifar data, let rank 0 download first
+    torch.distributed.barrier()
+
+trainset = torchvision.datasets.CIFAR10(root='./data', train=True, download=True, transform=transform)
+
+if torch.distributed.get_rank() == 0:
+    # cifar data is downloaded, indicate other ranks can proceed
+    torch.distributed.barrier()
+
+trainloader = torch.utils.data.DataLoader(trainset, batch_size=16, shuffle=True, num_workers=2)
+
+testset = torchvision.datasets.CIFAR10(root='./data', train=False, download=True, transform=transform)
+testloader = torch.utils.data.DataLoader(testset, batch_size=4, shuffle=False, num_workers=2)
+
+classes = ('plane', 'car', 'bird', 'cat', 'deer', 'dog', 'frog', 'horse', 'ship', 'truck')
+
+########################################################################
+# Let us show some of the training images, for fun.
+
+# functions to show an image
+
+
+def imshow(img):
+    """Show image."""
+    img = img / 2 + 0.5    # unnormalize
+    npimg = img.numpy()
+    plt.imshow(np.transpose(npimg, (1, 2, 0)))
+    plt.show()
+
+
+# get some random training images
+dataiter = iter(trainloader)
+images, labels = next(dataiter)
+
+# show images
+imshow(torchvision.utils.make_grid(images))
+# print labels
+print(' '.join('%5s' % classes[labels[j]] for j in range(4)))
+
+########################################################################
+# 2. Define a Convolutional Neural Network
+# ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+# Copy the neural network from the Neural Networks section before and modify it to
+# take 3-channel images (instead of 1-channel images as it was defined).
+
+args = add_argument()
+
+
+class FP8Block(nn.Module):
+    """Transformer Block using TransformerEngine."""
+    def __init__(
+        self,
+        dim,
+        num_heads,
+        mlp_ratio=4.,
+        qkv_bias=False,
+        qk_scale=None,
+        drop=0.,
+        attn_drop=0.,
+        drop_path=0.,
+        act_layer=nn.GELU,
+        norm_layer=nn.LayerNorm
+    ):
+        """Constructor."""
+        super().__init__()
+        assert dim % 16 == 0
+        mlp_hidden_dim = int(dim * mlp_ratio)
+        init_method = partial(trunc_normal_, std=.02)
+        self.m = te.TransformerLayer(
+            dim,
+            mlp_hidden_dim,
+            num_heads,
+            hidden_dropout=drop,
+            attention_dropout=attn_drop,
+            self_attn_mask_type='padding',
+            layer_type='encoder',
+            init_method=init_method,
+            output_layer_init_method=init_method,
+            drop_path_rate=drop_path,
+            fuse_qkv_params=True,
+        )
+
+    def forward(self, x):
+        """Forward computation."""
+        # x: seq_len, batch_size, dim
```

Comment:
The code L151-159 can be simplified with `return self.m(x, attention_mask=None)`

Reply:
Without padding, it will crash.
![image](https://github.com/Azure/MS-AMP/assets/4133758/6902987e-5a65-47be-a08b-5c052ebd9d46)